File: gfg/POTD/2025-02-04/solution.java

/*
class Node {
    int data;
    Node left;
    Node right;
    Node(int data) {
        this.data = data;
        left = null;
        right = null;
    }
}
*/

class Solution {
    int maxDiameter = 0;

    int diameter(Node root) {
        calculateDepth(root);
        return maxDiameter;
    }

    int calculateDepth(Node root) {
        if (root == null) return 0;

        int leftDepth = calculateDepth(root.left);
        int rightDepth = calculateDepth(root.right);

        maxDiameter = Math.max(maxDiameter, leftDepth + rightDepth);

        return 1 + Math.max(leftDepth, rightDepth);
    }
}

File: gfg/POTD/2025-02-04/solution.js
/*
class Node{
    constructor(data){
        this.data = data;
        this.left = null;
        this.right = null;
    }
}
*/

class Solution {
	/**
	 * @param {Node} root
	 * @returns {number}
	 */
	diameter(root) {
		let maxDiameter = 0

		const calculateDepth = (root) => {
			if (root === null) return 0
			const leftDepth = calculateDepth(root.left)
			const rightDepth = calculateDepth(root.right)

			maxDiameter = Math.max(maxDiameter, leftDepth + rightDepth)
			return 1 + Math.max(leftDepth, rightDepth)
		}
		calculateDepth(root)
		return maxDiameter
	}
}
